### PR TITLE
Add a friendly and easy-to-read UI for WPT tests

### PR DESCRIPTION
**Pull request by mrobinson**
The idea is that this UI will be installed adhoc by the Servo scripts
for now. Later, after baking for a while in the Servo source tree it
will be upstreamed to the mozlog project itself.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8141)

### CODE REVIEW

**Comment by mrobinson** [comment]
Here's a gif which demonstrates the new UI.

new-ui

cc @jgraham @jdm

**Comment by nox** [comment]
I don't see stacktraces and whatnot anymore, is that intended? How do I understand why a test failed with this?

**Comment by jdm** [comment]
While I like the output for the common case, like @nox I'm concerned about being able to extract useful output from tests that only fail intermittently or when run in a group.

**Comment by jdm** [comment]
Also, what's the output like for non-reftests? There can be failures for individual subtests as well as the overall test, and that information is valuable too.

**Comment by glennw** [comment]
I really like this UI for local development where I often want to run the tests quite iteratively. I find the existing output really hard to parse during development. Even if we can't use it for build machines, perhaps we could have it as an option.

**Comment by jgraham** [comment]
Sure, making it an option is easy. Making it the default is more controversial. I agree that some improvements to the output could be made; I don't think most of the requests here are that complicated?

**Comment by mrobinson** [comment]
I'm working on a more verbose version which, I hope, addresses all of the feedback, but maintains the readability of the output. I hope to have the updated PR available today.

**Comment by mrobinson** [comment]
I've updated the PR so that the output is more verbose. It now includes subtest failure messages and the most relevant line of the stack trace (if one exists). Here's how it looks now:

new-ui

**Comment by jdm** [comment]
:heart_eyes:

**Comment by jgraham** [comment]
Review status: 0 of 2 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[tests/wpt/friendly_formatter.py, line 9 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JGe_fY-zTcHwLso25)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L9)):</sup>
I don't really like the name. InteractiveTerminalFormatter, perhaps?


---

<sup>**[tests/wpt/friendly_formatter.py, line 42 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JVKigIKejMOrUCMUN)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L42)):</sup>
This doesn't seem very portable. In particular I think it doesn't work on Windows. Should we perhaps use something like Blessings instead that provides a higher level API (might still need colorama for Windows support)? Not a blocker for this PR though.


---

<sup>**[tests/wpt/friendly_formatter.py, line 62 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JH0kF2_cMRROgyZt5)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L62)):</sup>
Note that this can be 0 when we don't know how many tests will run.


---

<sup>**[tests/wpt/friendly_formatter.py, line 90 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JYepwHoVv9XEgIojL)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L90)):</sup>
If the overall test failed shouldn't this contain the message and stack (this can happen if a js test throws an exception not associated with any particular subtest, for example)


---

<sup>**[tests/wpt/friendly_formatter.py, line 110 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JXMPJAOZMh3gi83ka)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L110)):</sup>
Avoid line continuation characters throughout


---

<sup>**[tests/wpt/friendly_formatter.py, line 119 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JeLbWDZYXzSuRhp7C)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L119)):</sup>
Why don't we output extra data like the message and stack here?


---

<sup>**[tests/wpt/friendly_formatter.py, line 132 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JePybnPz4K87lSYf5)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L132)):</sup>
Also details of the subtest here.


---

<sup>**[tests/wpt/run.py, line 30 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JGI4SRND6jMYXHqcz)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/run.py#L30)):</sup>
Don't use line continuation characters. It's fine to just open the paren on one line and continue on the next.


---

<sup>**[tests/wpt/run.py, line 32 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JG4to_qpfItnpmewX)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/run.py#L32)):</sup>
Let's not change the default for now, not least because I doubt this will work well on buildbot, so we'll want to explicitly choose something else there first.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8141)

**Comment by nox** [comment]
I really want to keep displaying the (JS or Rust) stacktraces when a test fail.

**Comment by jgraham** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8141)

**Comment by jgraham** [comment]
Right, to be clear, my preferred model would be one-line summaries when the test initially fails (like now) and then the full stack and error message printed at the end of the test run.

**Comment by mrobinson** [comment]
@nox the stack is still displayed, but only the top line. Do you find yourself using more than that?

**Comment by mrobinson** [comment]
@jgraham I think I'd rather have the details in the scrollback, because the end is supposed to be a summary. There is also a practical reason as well. When a test fails in the middle of a run, it becomes possible to debug it while the run is finishing. You don't have to wait until the end. 

Additionally, detailed information about subtests is difficult to display in the summary since each subtest can fail for a different reason. Subtests names are harder to scan than file paths and I think it would make the summary a bit too complicated.

**Comment by jgraham** [comment]
I certainly use the detailed information including more than just the top line of the stacktrace.

I agree that having some output whilst the tests are running is an improvement. I'd just like to also see more verbose output in the final summary. There is precedence for this in many other test harnesses (e.g. python's unittest module and py.test), so I don't think the complexity of the output outweighs the usefulness of having the data.

Oh and whilst I think of it, what happens if servo crashes during a test run? Will you get the rust backtrace (assuming RUST_BACKTRACE=1, which I think is not yet the default). I suspect you won't.

**Comment by mrobinson** [comment]
The latest version makes a few changes:
   * The full stacktrace of subtest failures is included as soon as a test finishes. Subtests that fail with identical stracktraces are grouped together to avoid flooding the terminal.
   * When Servo itself crashes the Rust backtrace is included in the output.
   * The failing tests are no longer listed in the summary section, which instead only prints test counts now.
   * When the formatter isn't running on an interactive terminal, the constantly updating display is not printed or erased. Instead only test failures and the final summary are. This should make the formatter useful for redirection and running on the build bots.

Here's an example of a completed test run with variety of failures:
```
Running 4 tests in web-platform-tests

  ▶ Unexpected subtest result in /url/a-element.html:
  │ FAIL [expected PASS] Parsing: <http://example\t.\norg> against <http://example.org/foo/bar>
  │   → assert_equals: scheme expected (string) "http:" but got (undefined) undefined
  │ FAIL [expected PASS] Parsing: <http://user:pass@foo:21/bar;par?b#c> against <http://example.org/foo/bar>
  │   → assert_equals: scheme expected (string) "http:" but got (undefined) undefined
  │ FAIL [expected PASS] Parsing: <http:foo.com> against <http://example.org/foo/bar>
  │   → assert_equals: scheme expected (string) "http:" but got (undefined) undefined
  │ FAIL [expected PASS] Parsing: <\t   :foo.com   \n> against <http://example.org/foo/bar>
  │   → assert_equals: scheme expected (string) "http:" but got (undefined) undefined
  │ FAIL [expected PASS] Parsing: < foo.com  > against <http://example.org/foo/bar>
  │   → assert_equals: scheme expected (string) "http:" but got (undefined) undefined
  │ FAIL [expected PASS] Parsing: <a:\t foo.com> against <http://example.org/foo/bar>
  │   → assert_equals: scheme expected (string) "a:" but got (undefined) undefined
  │ 
  │ runURLTests/<@http://web-platform.test:8000/url/a-element.html:33:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1379:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  │ runURLTests@http://web-platform.test:8000/url/a-element.html:29:1
  │ request.onload<@http://web-platform.test:8000/url/a-element.html:8:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1379:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1403:1

  ▶ Unexpected subtest result in /url/a-element.html:
  │ FAIL [expected PASS] Loading data\xe2\x80\xa6
  │   → assert_equals: colors expected "green" but got "blue"
  │ 
  │ runURLTests@http://web-platform.test:8000/url/a-element.html:42:3
  │ request.onload<@http://web-platform.test:8000/url/a-element.html:8:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1379:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1403:1

  ▶ TIMEOUT [expected OK] /url/url-constructor.html

  ▶ Unexpected subtest result in /url/url-constructor.html:
  └ NOTRUN [expected PASS] Untitled

Ran 4 tests finished in 14.0 seconds.
  • 2 ran as expected. 0 tests skipped.
  • 1 tests timed out unexpectedly
  • 2 tests had unexpected subtest results
```

Here's an example of tests that fail with Servo crashes:
```
Running 4 tests in web-platform-tests

  ▶ CRASH [expected OK] /url/interfaces.html
  │ 
  │ thread '<main>' panicked at 'HOLY CRAP!', /home/martin/work/mozilla/servo/components/util/opts.rs:448
  │ stack backtrace:
  │    1:     0x7f37c39c1e50 - sys::backtrace::tracing::imp::write::h23bfa7a49ed99d87Klt
  │    2:     0x7f37c39c43a5 - panicking::log_panic::_<closure>::closure.39180
  │    3:     0x7f37c39c3df1 - panicking::log_panic::h846e3da4ec5bd0b1Wlx
  │    4:     0x7f37c39b0eb3 - sys_common::unwind::begin_unwind_inner::h04ad996baf6f81c7bds
  │    5:     0x7f37c2e6d717 - sys_common::unwind::begin_unwind::begin_unwind::h12814339548133697019
  │                         at src/libstd/sys/common/unwind/mod.rs:214
  │    6:     0x7f37c2e73b02 - opts::from_cmdline_args::h0c34c94b18057eb7hse
  │                         at /home/martin/work/mozilla/servo/components/servo/<std macros>:3
  │    7:     0x7f37c11070b1 - main::h8fd21a89dceb8751naa
  │                         at /home/martin/work/mozilla/servo/components/servo/main.rs:42
  │    8:     0x7f37c39c3b34 - sys_common::unwind::try::try_fn::h10604132475359592391
  │    9:     0x7f37c39c1388 - __rust_try
  │   10:     0x7f37c39c37cf - rt::lang_start::hd74bfb7a606308914ix
  │   11:     0x7f37c11df489 - main
  │   12:     0x7f37bebb1a3f - __libc_start_main
  │   13:     0x7f37c1106f28 - _start
  └   14:                0x0 - <unknown>

  ▶ CRASH [expected OK] /url/url-constructor.html
  │ 
  │ thread '<main>' panicked at 'HOLY CRAP!', /home/martin/work/mozilla/servo/components/util/opts.rs:448
  │ stack backtrace:
  │    1:     0x7fc2e7420e50 - sys::backtrace::tracing::imp::write::h23bfa7a49ed99d87Klt
  │    2:     0x7fc2e74233a5 - panicking::log_panic::_<closure>::closure.39180
  │    3:     0x7fc2e7422df1 - panicking::log_panic::h846e3da4ec5bd0b1Wlx
  │    4:     0x7fc2e740feb3 - sys_common::unwind::begin_unwind_inner::h04ad996baf6f81c7bds
  │    5:     0x7fc2e68cc717 - sys_common::unwind::begin_unwind::begin_unwind::h12814339548133697019
  │                         at src/libstd/sys/common/unwind/mod.rs:214
  │    6:     0x7fc2e68d2b02 - opts::from_cmdline_args::h0c34c94b18057eb7hse
  │                         at /home/martin/work/mozilla/servo/components/servo/<std macros>:3
  │    7:     0x7fc2e4b660b1 - main::h8fd21a89dceb8751naa
  │                         at /home/martin/work/mozilla/servo/components/servo/main.rs:42
  │    8:     0x7fc2e7422b34 - sys_common::unwind::try::try_fn::h10604132475359592391
  │    9:     0x7fc2e7420388 - __rust_try
  │   10:     0x7fc2e74227cf - rt::lang_start::hd74bfb7a606308914ix
  │   11:     0x7fc2e4c3e489 - main
  │   12:     0x7fc2e2610a3f - __libc_start_main
  │   13:     0x7fc2e4b65f28 - _start
  └   14:                0x0 - <unknown>

  ▶ CRASH [expected OK] /url/a-element.html
  │ 
  │ thread '<main>' panicked at 'HOLY CRAP!', /home/martin/work/mozilla/servo/components/util/opts.rs:448
  │ stack backtrace:
  │    1:     0x7fb9e9d13e50 - sys::backtrace::tracing::imp::write::h23bfa7a49ed99d87Klt
  │    2:     0x7fb9e9d163a5 - panicking::log_panic::_<closure>::closure.39180
  │    3:     0x7fb9e9d15df1 - panicking::log_panic::h846e3da4ec5bd0b1Wlx
  │    4:     0x7fb9e9d02eb3 - sys_common::unwind::begin_unwind_inner::h04ad996baf6f81c7bds
  │    5:     0x7fb9e91bf717 - sys_common::unwind::begin_unwind::begin_unwind::h12814339548133697019
  │                         at src/libstd/sys/common/unwind/mod.rs:214
  │    6:     0x7fb9e91c5b02 - opts::from_cmdline_args::h0c34c94b18057eb7hse
  │                         at /home/martin/work/mozilla/servo/components/servo/<std macros>:3
  │    7:     0x7fb9e74590b1 - main::h8fd21a89dceb8751naa
  │                         at /home/martin/work/mozilla/servo/components/servo/main.rs:42
  │    8:     0x7fb9e9d15b34 - sys_common::unwind::try::try_fn::h10604132475359592391
  │    9:     0x7fb9e9d13388 - __rust_try
  │   10:     0x7fb9e9d157cf - rt::lang_start::hd74bfb7a606308914ix
  │   11:     0x7fb9e7531489 - main
  │   12:     0x7fb9e4f03a3f - __libc_start_main
  │   13:     0x7fb9e7458f28 - _start
  └   14:                0x0 - <unknown>

  ▶ CRASH [expected TIMEOUT] /url/a-element.xhtml
  │ 
  │ thread '<main>' panicked at 'HOLY CRAP!', /home/martin/work/mozilla/servo/components/util/opts.rs:448
  │ stack backtrace:
  │    1:     0x7f0237810e50 - sys::backtrace::tracing::imp::write::h23bfa7a49ed99d87Klt
  │    2:     0x7f02378133a5 - panicking::log_panic::_<closure>::closure.39180
  │    3:     0x7f0237812df1 - panicking::log_panic::h846e3da4ec5bd0b1Wlx
  │    4:     0x7f02377ffeb3 - sys_common::unwind::begin_unwind_inner::h04ad996baf6f81c7bds
  │    5:     0x7f0236cbc717 - sys_common::unwind::begin_unwind::begin_unwind::h12814339548133697019
  │                         at src/libstd/sys/common/unwind/mod.rs:214
  │    6:     0x7f0236cc2b02 - opts::from_cmdline_args::h0c34c94b18057eb7hse
  │                         at /home/martin/work/mozilla/servo/components/servo/<std macros>:3
  │    7:     0x7f0234f560b1 - main::h8fd21a89dceb8751naa
  │                         at /home/martin/work/mozilla/servo/components/servo/main.rs:42
  │    8:     0x7f0237812b34 - sys_common::unwind::try::try_fn::h10604132475359592391
  │    9:     0x7f0237810388 - __rust_try
  │   10:     0x7f02378127cf - rt::lang_start::hd74bfb7a606308914ix
  │   11:     0x7f023502e489 - main
  │   12:     0x7f0232a00a3f - __libc_start_main
  │   13:     0x7f0234f55f28 - _start
  └   14:                0x0 - <unknown>

Ran 4 tests finished in 2.0 seconds.
  • 0 ran as expected. 0 tests skipped.
  • 4 tests crashed unexpectedly
```

**Comment by mrobinson** [comment]
Review status: 0 of 2 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

<sup>**[tests/wpt/friendly_formatter.py, line 9 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JGe_fY-zTcHwLso25)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L9)):</sup>
I've changed this to GroupingFormatter, because it now works on non-interactive terminals as well. I'm not super happy with the name, so feel free to suggest something else.


---

<sup>**[tests/wpt/friendly_formatter.py, line 42 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JVKigIKejMOrUCMUN)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L42)):</sup>
It doesn't look like blessings works with Windows. https://github.com/erikrose/blessings/issues/21. Possibly I can detect Windows explicitly at a future date and output the appropriate characters.


---

<sup>**[tests/wpt/friendly_formatter.py, line 62 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JH0kF2_cMRROgyZt5)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L62)):</sup>
I handle this situation explicitly now.


---

<sup>**[tests/wpt/friendly_formatter.py, line 90 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JYepwHoVv9XEgIojL)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L90)):</sup>
This works as expected now.


---

<sup>**[tests/wpt/friendly_formatter.py, line 110 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JXMPJAOZMh3gi83ka)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L110)):</sup>
I think I've eliminated all of these.


---

<sup>**[tests/wpt/friendly_formatter.py, line 119 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JeLbWDZYXzSuRhp7C)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L119)):</sup>
This is in the scrollback now.


---

<sup>**[tests/wpt/friendly_formatter.py, line 132 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JePybnPz4K87lSYf5)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L132)):</sup>
Ditto.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8141)

**Comment by mrobinson** [comment]
Review status: 0 of 2 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

<sup>**[tests/wpt/run.py, line 32 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JG4to_qpfItnpmewX)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/run.py#L32)):</sup>
I am still changing the default since this should work fine on the buildbot. I think it is worth trying there since we always have the full log as well.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8141)

**Comment by jgraham** [comment]
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[tests/wpt/grouping_formatter.py, line 47 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BiH1LT4MplfFDZ18p)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L47)):</sup>
Please file a followup issue about Windows support.


---

<sup>**[tests/wpt/grouping_formatter.py, line 68 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BizKWBpqli_u0TdfZ)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L68)):</sup>
Please don't use line continuation characters


---

<sup>**[tests/wpt/grouping_formatter.py, line 102 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BpsLdJVUXkjnJ_ZtC)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L102)):</sup>
This will do horrible things to non-ascii non-control characters, I think. Can we just have a fixed list of characters to replace with escapes or similar?


---

<sup>**[tests/wpt/grouping_formatter.py, line 134 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BtKeNKG6b_ifC7mvv)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L134)):</sup>
Doesn't calling this on self work? Otherwise just make it a normal method that doesn't use self.


---

<sup>**[tests/wpt/grouping_formatter.py, line 233 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2Bughgj_IO72epjH7z)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L233)):</sup>
I wonder if we want all log messages with some level like warning or above to be displayed.


---

<sup>**[tests/wpt/friendly_formatter.py, line 9 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JGe_fY-zTcHwLso25)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L9)):</sup>
We can think of something better when we upstream this.


---

<sup>**[tests/wpt/friendly_formatter.py, line 42 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JVKigIKejMOrUCMUN)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L42)):</sup>
I think colorama is actually what we need here, but this can be defered,


---

<sup>**[tests/wpt/friendly_formatter.py, line 119 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8141#-K1JeLbWDZYXzSuRhp7C)** ([raw file](https://github.com/servo/servo/blob/5c868d9ea5648796b3de2302e2d8d7a30d3d74ed/tests/wpt/friendly_formatter.py#L119)):</sup>
(FYI I'm happy for you to click "acknowledge" if you fixed an issue)


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8141)

**Comment by mrobinson** [comment]
Review status: 1 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[tests/wpt/grouping_formatter.py, line 47 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BiH1LT4MplfFDZ18p)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L47)):</sup>
Done.


---

<sup>**[tests/wpt/grouping_formatter.py, line 68 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BizKWBpqli_u0TdfZ)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L68)):</sup>
Done.


---

<sup>**[tests/wpt/grouping_formatter.py, line 102 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BpsLdJVUXkjnJ_ZtC)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L102)):</sup>
The test name is a Unicode Python string, and I am encoding here, not decoding. I don't understand why UTF-8 encoding will affect non-ASCII characters. Could you expand a bit on the problem that you see?


---

<sup>**[tests/wpt/grouping_formatter.py, line 134 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BtKeNKG6b_ifC7mvv)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L134)):</sup>
Done.


---

<sup>**[tests/wpt/grouping_formatter.py, line 233 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2Bughgj_IO72epjH7z)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L233)):</sup>
I can do this for ERROR and CRITICAL, but the test runner prints many messages such as "runner teardown" with the WARNING level as part of normal operation.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8141)

**Comment by jgraham** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[tests/wpt/grouping_formatter.py, line 102 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BpsLdJVUXkjnJ_ZtC)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L102)):</sup>
    In [3]: print u"\tabc\u0234"
    	abcȴ

    In [5]: print u"\tabc\u0234".encode("utf8").encode("string-escape")
    \tabc\xc8\xb4


---

<sup>**[tests/wpt/grouping_formatter.py, line 233 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2Bughgj_IO72epjH7z)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L233)):</sup>
OK, let's do that for ERROR and CRITICAL then, and maybe file a bug to make the others INFO.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8141)

**Comment by mrobinson** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[tests/wpt/grouping_formatter.py, line 102 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BpsLdJVUXkjnJ_ZtC)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L102)):</sup>
Done. I'm using unicode-escape now. This will still convert inline non-ASCII characters into \uxxx sequences, but I think this is better than trying to fix every case that appears. For instance see tests/wpt/web-platform-tests/url/urltestdata.txt.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8141)

**Comment by jgraham** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---

<sup>**[tests/wpt/grouping_formatter.py, line 102 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8141#-K2BpsLdJVUXkjnJ_ZtC)** ([raw file](https://github.com/servo/servo/blob/f74c855f3b22b8e78343a045cf8ea4181751aeb7/tests/wpt/grouping_formatter.py#L102)):</sup>
I still disagree and think we want to avoid escapes where possible, but we can fix this later.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8141)

**Comment by jgraham** [comment]
r+ (someone will have to tell homu)
